codegen-v2/src/main/kotlin/club/doki7/babel/cdecl/parse.kt

```diff
@@ -1,11 +1,424 @@
+                }
+                else -> {
+                    break
+                }
+            }
+            pos++
+        }
+        if (typeName == null) {
+            throw ParseException("expected type name", curToken)
+        }
+        return RawIdentifierType(typeName, modifiers, trivia)
+    }
+
+    // prefix-declarator :: '*' type-qualifier*
+    private fun parsePrefixDeclarator(): RawType? {
+        var trivia = mutableListOf<String>()
+        var ty: RawPointerType? = null
+        while (true) {
+            trivia.addAll(getAndSkipTrivia())
+            if (hasEnded) {
+                break
+            }
+            when (curToken.type) {
+                TokenType.SYMBOL -> {
+                    if (curToken.value == "*") {
+                        ty = if (ty == null) {
+                            RawPointerType(EmptyType(), const = false, volatile = false, syntaxTrivia = trivia)
+                        } else {
+                            RawPointerType(ty, const = false, volatile = false, syntaxTrivia = trivia)
+                        }
+                        trivia = mutableListOf()
+                    } else {
+                        break
+                    }
+                }
+                TokenType.QUALIFIER -> {
+                    if (ty == null) {
+                        throw ParseException("expected pointer before qualifier", curToken)
+                    }
+                    when (curToken.value) {
+                        CONST -> ty.const = true
+                        VOLATILE -> ty.volatile = true
+                        else -> throw ParseException("unexpected qualifier", curToken)
+                    }
+                }
+                else -> break
+            }
+            pos++
+        }
+        return ty
+    }
+
+    private fun parseParameterList(): List<Pair<String, RawType>> {
+        val params = mutableListOf<Pair<String, RawType>>()
+        if (hasEnded) {
+            unexpectedEnd("expected ')' or parameter declaration")
+        }
+        while (true) {
+            if (curToken.type == TokenType.SYMBOL && curToken.value == ")") {
+                break
+            }
+            params.add(parseDeclaration())
+            if (hasEnded) {
+                unexpectedEnd("expected ',' or ')'")
+            } else {
+                if (curToken.type == TokenType.SYMBOL && curToken.value == ",") {
+                    pos++
+                } else if (curToken.type == TokenType.SYMBOL && curToken.value == ")") {
+                    break
+                } else {
+                    throw ParseException("expected ',' or ')'", curToken)
+                }
+            }
+        }
+        pos++
+        return params
+    }
+
+    private fun parseArrayLength(): String {
+        if (hasEnded) {
+            unexpectedEnd("expected array length")
+        }
+        if (curToken.type != TokenType.INTEGER) {
```

Comment:
Array length does not need to be an integer. Here is a tricky point, let me do it.